core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java

```diff
@@ -0,0 +1,220 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class JdbcTableOperations extends BaseMetastoreTableOperations {
+
+  private static final Logger LOG = LoggerFactory.getLogger(JdbcTableOperations.class);
+  private final String catalogName;
+  private final TableIdentifier tableIdentifier;
+  private final FileIO fileIO;
+  private final JdbcClientPool connections;
+
+  protected JdbcTableOperations(JdbcClientPool dbConnPool, FileIO fileIO, String catalogName,
+                                TableIdentifier tableIdentifier) {
+    this.catalogName = catalogName;
+    this.tableIdentifier = tableIdentifier;
+    this.fileIO = fileIO;
+    this.connections = dbConnPool;
+  }
+
+  @Override
+  public void doRefresh() {
+    Map<String, String> table;
+
+    try {
+      table = getTable();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new UncheckedInterruptedException(e, "Interrupted during refresh");
+    } catch (SQLException e) {
+      // SQL exception happened when getting table from catalog
+      throw new UncheckedSQLException(e, "Failed to get table %s from catalog %s", tableIdentifier, catalogName);
+    }
+
+    // Table not exists AND currentMetadataLocation is not NULL!
+    if (table.isEmpty() && currentMetadataLocation() != null) {
+      throw new NoSuchTableException("Failed to get table %s from catalog %s" +
+          " maybe another process deleted it", tableIdentifier, catalogName);
+    }
+
+    // Table not exists in the catalog! metadataLocation is null here!
+    if (table.isEmpty()) {
+      refreshFromMetadataLocation(null);
+      return;
+    }
+
+    // Table exists but metadataLocation is null
+    if (table.getOrDefault(JdbcUtil.METADATA_LOCATION, null) == null) {
+      throw new RuntimeException(String.format("Failed to get metadata location of the table %s from catalog %s",
+          tableIdentifier, catalogName));
+    }
+
+    refreshFromMetadataLocation(table.get(JdbcUtil.METADATA_LOCATION));
+  }
+
+  @Override
+  public void doCommit(TableMetadata base, TableMetadata metadata) {
+    String newMetadataLocation = writeNewMetadata(metadata, currentVersion() + 1);
+    try {
+      Map<String, String> table = getTable();
+
+      if (!table.isEmpty()) {
+        validateMetadataLocation(table, base);
+        String oldMetadataLocation = base.metadataFileLocation();
+        // Start atomic update
+        LOG.debug("Committing existing table: {}", tableName());
+        updateTable(newMetadataLocation, oldMetadataLocation);
+      } else {
+        // table not exists create it
+        LOG.debug("Committing new table: {}", tableName());
+        createTable(newMetadataLocation);
+      }
+
+    } catch (SQLIntegrityConstraintViolationException e) {
+      throw new AlreadyExistsException(e, "Table already exists, maybe another process created it");
```

Comment:
I think this should only be thrown if `currentMetadataLocation()` is `null` because that indicates this is a new table. Otherwise, I think this should throw `UncheckedSQLException`.

Reply:
updated with following if condition

```
      if (currentMetadataLocation() == null) {
        throw new AlreadyExistsException(e, "Table already exists: %s", tableIdentifier);
      } else {
        throw new UncheckedSQLException(e, "Table already exists: %s", tableIdentifier);
      }
```